src/crimson/os/seastore/cache.cc

```diff
@@ -1486,13 +1495,18 @@ void Cache::complete_commit(
     }
   }
 
-  for (auto &i: t.retired_set) {
-    epm.mark_space_free(i->get_paddr(), i->get_length());
+  // The range needs to be recalculated here not to free used space on 
+  // existing_block_list. If we free the overlapped space between
+  // retired_set and existing_block_list, another transaction may use
+  // the space before calling mark_space_used to the extent
+  // in existing_block_list.
```

Comment:
In RBM, block allocation is done before the transaction begins. Therefore, if we free the space via mark_space_free here, another block allocation request can use the freed block. Am I wrong?

Reply:
The space allocations from `existing_block_list` are atomically updated (freed and mark_used) in `complete_commit()`, so there is really no chance for another allocation request to be executed in between. And I think `mark_space_used()` should mark the space as used for RBM to correctly reflect the commited change that an original extent was remapped to smaller sub extent(s).

Reply:
Please consider the following scenario in test_transaction_manager.

1. dec_ref is called to range A
2. map_existing_extent() is called to a part of the range A,

The range A is added to retired_set, but a part of the range A is also added to existing_block_list. Therefore, if we free all chunks in retired_set---this free marks the corresponding regions in extent allocator as free---even though the part is included in the existing_block_list, in-memory block allocator (extent allocator) in RBM can allocate the part for another purpose because this is not reserved. This causes inconsistency because existing_block_list contains and use the part.

Reply:
> 1. dec_ref is called to range A
> 2. map_existing_extent() is called to a part of the range A

Hmm, I think the test case should be wrong. User should not remap an extent after retiring it -- it doesn't exist any more. Are you able to point out the specific test code?

Reply:
```
  void test_map_existing_extent() {
    run_async([this] {
      constexpr size_t offset = 16 << 10;
      constexpr size_t length = 16 << 10;
      {
        auto t = create_transaction();
        auto extent = alloc_extent(t, offset, length);
        submit_transaction(std::move(t));
      } 
      { 
        auto t = create_transaction();
        auto extent = get_extent(t, offset, length);
        auto base_paddr = extent->get_paddr();
        dec_ref(t, offset);
        auto extent1 = map_existing_extent(t, offset, base_paddr, 4 << 10);
        ASSERT_TRUE(extent1.has_value());
        auto extent2 = map_existing_extent(t, offset + (4 << 10), base_paddr.add_offset(4 << 10), 4 << 10); 
        ASSERT_TRUE(extent2.has_value());
        auto extent3 = map_existing_extent(t, offset + (8 << 10), base_paddr.add_offset(8 << 10), 8 << 10);
        ASSERT_TRUE(extent3.has_value());
        ASSERT_TRUE((*extent1)->is_exist_clean());
        ASSERT_TRUE((*extent2)->is_exist_clean());
        ASSERT_TRUE((*extent3)->is_exist_clean());
        auto extent4 = mutate_extent(t, (*extent3));
        ASSERT_TRUE(extent4->is_exist_mutation_pending());
        ASSERT_TRUE((*extent3).get() == extent4.get());
        submit_transaction(std::move(t));
        check();
      } 
      replay();
      check();
    }); 
  } 
```

Reply:
```
DEBUG 2022-11-11 16:49:16,289 [shard 0] seastore_tm - 0x617000404280 TransactionManager::dec_ref: extent refcount is decremented to 0 -- 16384~16384, paddr<Dev(128),67112960>
DEBUG 2022-11-11 16:49:16,289 [shard 0] seastore_cache - 0x617000404280 Cache::retire_extent_addr: retire paddr<Dev(128),67112960>~16384 on t -- CachedExtent(addr=0x61200034d140, type=TEST_BLOCK, version=0, dirty_from_or_retired_at=JOURNAL_SEQ_NULL, modify_time=tp(NULL), paddr=paddr<Dev(128),67112960>, length=16384, state=CLEAN, last_committed_crc=4150607820, refcount=4, user_hint=Hint(NULL), rewrite_gen=GEN_NULL, laddr=16384, pin=LBAPin(16384~16384->paddr<Dev(128),67112960>)
DEBUG 2022-11-11 16:49:16,289 [shard 0] test - dec_ref: 16384 at refcount 0
DEBUG 2022-11-11 16:49:16,289 [shard 0] seastore_tm - 0x617000404280 TransactionManager::map_existing_extent:  laddr_hint: 16384 existing_paddr: paddr<Dev(128),67112960> length: 4096
DEBUG 2022-11-11 16:49:16,289 [shard 0] seastore_lba - 0x617000404280 BtreeLBAManager::alloc_extent: paddr<Dev(128),67112960>~4096, hint=16384, state: end, done with 1 attempts, insert at 16384
```

Reply:
Ah, I forget the fact that user is responsible to retire the original extent for the remapping. `map_existing_extent()` won't do the retirement itself. The test is correct.

> Please consider the following scenario in test_transaction_manager.
> 
> 1. dec_ref is called to range A
> 2. map_existing_extent() is called to a part of the range A,
>
> The range A is added to retired_set, but a part of the range A is also added to existing_block_list. Therefore, if we free all chunks in retired_set---this free marks the corresponding regions in extent allocator as free---even though the part is included in the existing_block_list, in-memory block allocator (extent allocator) in RBM can allocate the part for another purpose because this is not reserved. This causes inconsistency because existing_block_list contains and use the part.

For the range A which is in the `retired_set` (to be freed) and the sub-range being remapped in the `existing_block_list` (to be used again), the following logic is atomic in `Cache::complete_commit()`: https://github.com/ceph/ceph/blob/d5045f7142cb6edf496502ba85b25da78e249bf9/src/crimson/os/seastore/cache.cc#L1490-L1497

This atomicity is guaranteed as seastar only allocates one thread per shard, all the logics in the same shard can only be scheduled one after the other -- they just cannot be running in parallel in the same thread. In another word, it is impossible for the RBM allocator to have the chance to do any other allocations between these atomic freeing `retired_set` and using `existing_block_list` logics above. And there are a number of other places relying on this fact in order to be lockless.

Is there anything still missing?

Reply:
Understood.